Feature/add simulator api skeleton

efdx opened the pull request:
# Summary

In order to progress the use of simulator, implement the first revision / skeleton of the simulator ReST API. This should be usable for minimal usage of running a simulator and manipulating registers or causing modbus error responses.

This PR is made with the intention that the added API does not change the simulator, and instead tries to work on the existing base as much as possible. This is done so as not to push too many changes, making this PR intrusive or have unintended changes.

Add tests and documentation for later use and hopefully provide more detailed context to this PR.

Just to clarify, this PR in itself might not be release worthy in any ways. It is usable, but the choices made here are most likely not suitable to all use-cases. The API should be set in stone (possibly during this PR) prior to being advertised as complete.

Relates to issue #1284.

# Testing

1. Ran pytests, new tests pass and no regressions
2. The added API worked for own use-cases.

janiversen commented:
Seems you have some pylint error, please fix.

The easiest way is to have a local development (see the README), which will run pylon locally and even fix the majority of problems automatically.

efdx commented:
Will fix the CI errors.

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -199,7 +200,7 @@ def __init__(
             "calls_json": [None, self.build_json_calls],
```
not _json

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -199,7 +200,7 @@ def __init__(
             "calls_json": [None, self.build_json_calls],
             "server_json": [None, self.build_json_server],
```
not _json

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -301,14 +302,18 @@ async def handle_html(self, request):
-        json_dict = self.generator_html[page_type][0].copy()
-        result = self.generator_json[page_type][1](params, json_dict)
-        return web.Response(text=f"json build: {page_type} - {params} - {result}")
+        params = await request.json()
```
this looks very complicated, is sounds as if removing _json could simplify this a lot.

efdx replied:
From your immediate view, I get it, and kind of agree with it. From the point of anyone automating anything based on this, not so much. But please let me know if you plan to scrap it.

Also, I think many of the things could be streamlined with changing the structure of how these are handled. IMHO the only complexity comes from the current copy-paste model I adopted, as to not break anything else, and possibly to have this discussion. Changing the logic to have the data and encoding separate, preferrably with automatic serialization with json could help alot. But I felt it would have warranted too much changes to the existing and been too intrusive.

janiversen replied:
It is normal to have multiple endpoints for different types of service, so we clearly want 1 endpoint for html and another for RestAPI.

We do NOT want to use the same endpoint for different services and distinguish in the naming !

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -301,14 +302,18 @@ async def handle_html(self, request):
-        result = self.generator_json[page_type][1](params, json_dict)
-        return web.Response(text=f"json build: {page_type} - {params} - {result}")
+        params = await request.json()
+        json_dict = copy.deepcopy(self.generator_json[page_type][0])
```
what happened in page_type is unknown ??

the name "page_type" does not relate well to a RestApi request, did you mean "request"

efdx replied:
> what happened in page_type is unknown ??

The aiohttp handles the unregistered endpoint. Similar to what happens when using html.

> the name "page_type" does not relate well to a RestApi request, did you mean "request"

Did not pay any mind to it. I'll change it.

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
+        return json_response
 
-    def build_json_log(self, params, json_dict):
+    def build_json_log(self, _params, _json_dict):
```
If the parameters are unused, then why demand them ?

janiversen replied:
if a parameter is not needed, and there are no logic making the code simpler if its made dummy, then it should be removed.

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
+    def build_json_log(self, _params, _json_dict):
         """Build json log page."""
-        return f"json build log: {params} - {json_dict}"
+        return {"result": "error", "error": "log endpoint not implemented"}
```
ditto.

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -460,23 +465,154 @@ def build_html_server(self, _params, html):
+        return {"result": "error", "error": "log endpoint not implemented"}
 
-    def build_json_server(self, params, json_dict):
+    def build_json_server(self, _params, _json_dict):
```
ditto.

janiversen commented on pyproject.toml:
```diff
@@ -71,6 +71,7 @@ development = [
     "pytest-profiling>=1.7.0",
     "pytest-timeout>=2.3.1",
     "pytest-xdist>=3.6.1",
+    "pytest-aiohttp>=1.0.5",
```
Why is that needed now ?

it was not needed for the html ?

janiversen replied:
that might be.

janiversen commented on test/sub_server/test_simulator_api.py:
```diff
@@ -0,0 +1,467 @@
+"""Test simulator API."""
```
This is not right !! it is as if you have rewritten the whole file.

efdx replied:
What do you mean? It's a new file. Did you

1. Confuse it with test_simulator.py?
2. Want me to combine it with it?

janiversen replied:
Seems I got confused.

efdx commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -180,7 +181,7 @@ def __init__(
             [
                 web.get("/api/{tail:[a-z]*}", self.handle_html),
-                web.post("/api/{tail:[a-z]*}", self.handle_json),
+                web.post("/api/{tail:[a-z]*}_json", self.handle_json),
```
Sure. I'll change it.

janiversen replied:
Still not changed.

Esco441-91 replied:
Tomorrow. Ran out of hours in my day 🙂 22:39 here.

Esco441-91 replied:
(Aplogies for two accounts. Github doesn't seem to like my linking)

janiversen commented on doc/source/library/simulator/restapi.rst:
```diff
@@ -1,4 +1,120 @@
+
+The API provides the following endpoints:
+
+- /api/registers_json
```
not _json

janiversen reviewed (changes requested):
Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a lot to do, so I will not review the same code X times, just to see its unchanged.

Once you have addressed all comments, then click the "ready for review" button.

efdx commented:
> Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a > lot to do, so I will not review the same code X times, just to see its unchanged.
>
> Once you have addressed all comments, then click the "ready for review" button.

I am very confused over what you meant. I didn't ping that the requested changes are ready, the only thing I did yesterday evening was the comments/clarification requests on the opened discussions.

Was it perhaps the prior comment related to CI changes that came in your e-mail post-fact?

![image](https://github.com/user-attachments/assets/fcd3c6ab-2092-4921-af1d-0548680d6c18)

This one was before there were any change requests posted.

Please let me know.

EDIT: Need to fix CI stuff again.

efdx commented:
The CI checks now pass and I've changed the requested. Please let me know if anything else requires amending.

janiversen commented:
> > Please do not ping me, and say you changed all, when I see there are still quite a number of open comments. I do have quite a > lot to do, so I will not review the same code X times, just to see its unchanged.
> > Once you have addressed all comments, then click the "ready for review" button.
> 
> I am very confused over what you meant. I didn't ping that the requested changes are ready, the only thing I did yesterday evening was the comments/clarification requests on the opened discussions.
> 
> Was it perhaps the prior comment related to CI changes that came in your e-mail post-fact?
> 
> ![image](https://private-user-images.githubusercontent.com/150683046/354488603-fcd3c6ab-2092-4921-af1d-0548680d6c18.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1OTAyMDgsIm5iZiI6MTcyMjU4OTkwOCwicGF0aCI6Ii8xNTA2ODMwNDYvMzU0NDg4NjAzLWZjZDNjNmFiLTIwOTItNDkyMS1hZjFkLTA1NDg2ODBkNmMxOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgwMlQwOTExNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYjNkYWI0OGQ1MDU3MGFiMTg0ZjY1ZGI5MjIzMDk1NWIwNGRhNGFkY2MyYzQ5YzkyZTM0NWQzZjQ0ZTBjNTcxJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.lKSqtUNSxlEsdLVoyS0s7uEKcQShhBN6Bkopqt-CB5I)
> 
> This one was before there were any change requests posted.

Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -300,15 +300,18 @@ async def handle_html(self, request):
+        endpoint = request.path.split("/")[-1]
+        params = await request.json()
+        try:
+            result = self.generator_json[endpoint](params)
```
There are only one "endpoint" == restapi, which you defined earlier...I think you are mixing endpoints and commands.

efdx replied:
Yes. I often get confused with the terminology.

I'll change it to `command =`. You also had a previous suggestion on using "request" (which I missed/forgot). Let me know if you prefer that one over command.

efdx replied:
Changed

janiversen commented on pymodbus/server/simulator/http_server.py:
```diff
@@ -300,15 +300,18 @@ async def handle_html(self, request):
+        params = await request.json()
+        try:
+            result = self.generator_json[endpoint](params)
+        except Exception as exc: # pylint: disable=broad-exception-caught
```
We do not like broad exceptions, especially not when it is easy replaced by a strict exception like here.

efdx replied:
True, and my implementation is against best practices.

However, this one catches unhandled situations. Something like failed int conversion for "range_start/stop", or "value" when calling "Set". The same that aiohttp does by itself, except that it will post the data back in clear-text format, causing problems on Json parsing.

So I have few options:

1. Catch all the known exceptions in the existing code, either here on inplace with or without rethrowing as new, well-defined errors.
2. Leave it as is, handle them as unhandled
3. Combination. Catch all that are known, leave the broad exception in the event of any missed or future implementation errors.

Option 1 and 3 will require me to put in a partial improvement on json side as well, on top of everything.

Please let me know what you think.

efdx replied:
> on json side

on html side*

janiversen replied:
it is easy to see which exceptions can happen, so please limit it.

efdx replied:
Changed

efdx commented:
> Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.

I am still very confused. So for you, the post in the picture came AFTER you had reviewed my changes? If so, then I blame github for something going weird.

Regardless, I apologize if the post mentioned caused any confusion. I had no intention on pinging anyone without any changes done.

janiversen commented:
> > Actually it was NOT, I read your comment and started a review, until I realized that the majority of the items was not solved.
> 
> I am still very confused. So for you, the post in the picture came AFTER you had reviewed my changes for the first time? If so, then I blame github for something going weird.
> 
> Regardless, I apologize if the post mentioned caused any confusion. I had no intention on pinging anyone without any changes done.

No the post came AFTER my first review (17hours ago, and you responded to that with several comments), and BEFORE my second review, I just did the THIRD review.

Anyhow its history.

efdx commented:
Apologies for the long silence. Other work stuff got in the way, but it's a poor excuse.

What do you feel like about this PR, do you still want it in? I get that long drawn-out PRs are not very attractive to be considered merging. I can live with it.

For the record, I suggest checking out [commit a1a9864](https://github.com/pymodbus-dev/pymodbus/pull/2274/commits/a1a986407f0fc90700dbb5de5d375c575dfb989f) regardless from this topic. I figure extra output would be unnecessary? The output in question can be seen in any recent CI runs that run the pytest on ubuntu.

efdx commented:
I've now fixed the requested parts and put the PR up-to-date with the dev branch. Please let me know if you disagree or if there is anything else.

abaenaDSET commented:
> I've now fixed the requested parts and put the PR up-to-date with the dev branch. Please let me know if you disagree or if there is anything else.

I think you should close the old changes requested if they are already solved. Then the status of the PR will become pending to merge. I see you have left one discussion opened when you answered "Changed". After this I think you will be able to mark this PR as pending to resolve.

Thanks for your contributions :)

efdx commented:
> you should close the old changes requested if they are already solved. 

Thanks for the tips. I use mostly devops and gitlab internally, so I'm not that aware on all of the trick in github :)

However, I do not see a way I can mark the change request as "closed" myself. The best I can do is resolve each comment (tried), but this does not close the change request itself.

I think @janiversen needs to close them...?

janiversen commented:
The review comments are closed by hitting "resolved"

efdx commented:
> The review comments are closed by hitting "resolved"

The comments themselves I take it, but not the change request. Or then I'm misunderstanding something. If so, then sorry for the noise.

For me it shows everything is resolved, but it still shows "Changes requested" (and "Merging is blocked"). Same as mentioned in https://stackoverflow.com/questions/49780667/github-changes-requested-label-stay-after-i-answered-all-the-code-review-questi .

janiversen commented on doc/source/library/simulator/restapi.rst:
```diff
@@ -1,4 +1,120 @@
+This is still a Work In Progress. There may be large changes to the API in the
+future.
+
+There is a skeleton API for the Pymodbus simulator. The API is a simple copy of
```
```suggestion
There is a skeleton API for the Pymodbus simulator. The API is a simple copy of
```
The API is a simple copy of
```

janiversen reviewed (approved):
Lgtm, thanks.
